### PR TITLE
docs: revamp documentation

### LINKED ISSUE #528: Could you please extend https://awkward-array.org/how-to-use-in-numba-arraybuilder.html
Following the nice invite in the docs to raise an issue for unfilled topics that should get higher priority, I am doing that here.

I would like to write a numba function that returns an awkward array (specifically a ListOffsetArray) from other ListOffsetArrays. I need to run the equivalent of numpy.empty_like(...) and numpy.zeros_like(...) in numba-compiled Python with an awkward array. I suppose that filling the array is straight forward, I have trouble with the creation...

On a related note, is it possible to access the total size of a ListOffsetArray from within numba-compiled Python?

### LINKED ISSUE #753: How to use Awkward Arrays on GPUs
First of all, this is an amazing project! I am trying to use awkward arrays with CUDA kernels implemented with Numba. It would be good to have more information in general on how to use awkward arrays on GPU (on this page https://awkward-array.org/how-to-math-gpu.html) and, if possible, an example with Numba cuda-jitted function.

Thanks!

### LINKED ISSUE #1064: Awkward 1.0 Numba documentation
### Description of new feature

https://awkward-array.org/how-to-use-in-numba-features.html

### LINKED ISSUE #1139: How to use Awkward Arrays on GPUs
### Which documentation?

Python docstrings

### What needs to be documented?

I am wondering if I can use the awkward arrays on GPUs similar to dataframes or numpy arrays, (provided by rapids.ai and other tools). I came across this link
https://awkward-array.org/how-to-math-gpu.html

but it seems to be not in priority list. I just wanted to check if an implementation already exists. In case yes, an example will be really useful. 

Thanks.

### LINKED ISSUE #1144: Documentation
I looked up awkward array documentation (https://awkward-array.org/quickstart.html), but many are labeled [to do]. Can someone fill it up? For example, I tried to understand how to filter array but found the corresponding documentation missing. Many thanks.

### LINKED ISSUE #1366: Examining arrays documentation
### Which documentation?

Tutorials site

### What needs to be documented?

Just if such issue could prioritize writing some next part of the documentation. 

Always this arises every time uproot/awkward are used and every time it is a pain to find/remember

https://awkward-array.org/how-to-examine-type.html

### PR DESCRIPTION
**Pull request by jpivarski**
I'm generally giving this a once-over, and hopefully will fill in all of the TODO's.

But first, some general maintenance.

### CODE REVIEW

File: src/awkward/operations/ak_mean.py [hunk 1]
```diff
@@ -169,6 +169,7 @@ def nanmean(
         mask_identity,
         highlevel=highlevel,
         behavior=behavior,
+        attrs=attrs,
```
Comment (agoose77) [inline src/awkward/operations/ak_mean.py:172]: :heart_eyes:

**Comment by jpivarski** [comment]
Getting Started is done; User Guide is next.

**Comment by jpivarski** [comment]
Today, I filled in 6 of the 22 "[todo]" sections. The procedure is:

1. Selected user-guide pages from Plausible statistics with less than 80% bounce rate.
2. Sorted by total time on page (total pageviews × time on page).
3. Put the markdown for these pages, with most popular last, in the context window for ChatGPT.
4. Described the new page that needs to be written and let ChatGPT have a crack at it.[^1]
5. Open the result in Jupyter to run code cells.
6. Edit mercilessly.

There usually isn't much ChatGPT-generated content left at the end. Pointless adverbs are the first to go, but I end up replacing most paragraphs as well, and usually modify the code examples, too, at least a little bit.

So why use ChatGPT at all, if I'm going to be replacing everything it wrote?

Writer's block. Staring at a blank page, I find myself wondering what I had intended this stub to say 4 years ago. (These stubs were all created in PR #237, which was May 7, 2020. They've been waiting just a few days less than _4 whole years_ for me to fill them in!) When ChatGPT fills it in for me, I look at what it has to say and I get so disgusted by it that I replace it in little bits, like the ship of Theseus, with something halfway decent.

It's kind of like getting a coauthor to write their part of a paper by writing something terrible and asking, "How's this?"

[^1]: Interestingly, at one point, it copied an existing document almost verbatim (99% of the characters were the same), even though it has a rather high temperature: 1.0 (on a scale from 0 to 2, but 0.7 is most common). It doesn't _always_ generalize its input.

**Comment by jpivarski** [comment]
And it cost $4.53, shoving 28,000 words into `gpt-4-turbo` each time I tried to generate another zeroth draft. (It's well worth it, if means I can get all of this done in 2 or 3 more 7-hour sessions.)

**Comment by jpivarski** [comment]
The 1880 U.S. Census took 9 years and $5.8 million to analyze by hand. 

The 1890 Census, with the aid of Hollerith's electromechanical adding machines, took 7 years and $11.5 million to analyze, despite one-to-one tests showing that it was 8 to 10 times faster when employed on identical work.[^1]

What happened? The Census Bureau had fun with their new toy, performing calculations and analyzing the data in ways it has never been analyzed before. It's funny how quickly a labor-saving device becomes a labor-inspiring device.

[^1]: http://ds-wordpress.haverford.edu/bitbybit/bit-by-bit-contents/chapter-three/3-2-hollerith-and-the-census/

File: docs/user-guide/how-to-math-broadcasting.md [hunk 2]
```diff
@@ -14,10 +14,234 @@ kernelspec:
...
+
+* NumPy left-broadcasts,
+* Awkward Arrays with fixed-length lists left-broadcast, for consistency with NumPy,
+* Awkward Arrays with variable-length lists right-broadcast, for consistency with imperative code.
+
+One way to control this is to ensure that all arrays involved in an expression have the same number of dimensions by explicitly expanding them. Implicit broadcasting only happens for arrays of different numbers of dimensions, or if the length of a dimension is 1.
+
+But it might also be the case that your arrays have lists of equal length, so they seem to be regular like a NumPy array, yet their data type says that the lists _can_ be variable-length. Perhaps you got the NumPy-like data from a source that doesn't enforce fixed lengths, such as Python lists ({func}`ak.from_iter`), JSON ({func}`ak.from_json`), or Parquet ({func}`ak.from_parquet`). Check the array's {func}`ak.type` to see whether all dimensions are ragged (`var *`) or regular (some number `*`).
+
+The {func}`ak.from_regular` and {func}`ak.to_regular` functions toggle ragged (`var *`) and regular (some number `*`) dimensions.
```
Comment (agoose77) [inline docs/user-guide/how-to-math-broadcasting.md]: Also, `ak.enforce_type`! :)

File: docs/user-guide/how-to-math-broadcasting.md [hunk 3]
```diff
@@ -14,10 +14,234 @@ kernelspec:
...
+print(ak.Array(x).type)
+print(ak.Array(y).type)
+```
+
+In both NumPy and Awkward Array, `x` has fewer dimensions than `y`, so `x` is expanded on the _left_ from length-1 to length-2.
+
+However, if the Awkward Array has variable-length _type_, regardless of whether the actual lists have variable lengths,
+
+```{code-cell} ipython3
+print(ak.Array(x.tolist()).type)
```
Comment (agoose77) [inline docs/user-guide/how-to-math-broadcasting.md:167]: Should we prefer the `to_list` alias rather than `tolist`?

Comment (jpivarski) [reply docs/user-guide/how-to-math-broadcasting.md:167]: I'm torn on this for the same reasons that we have both aliases: NumPy's name for it is `tolist` and that's quicker to type, which matters in most of the instances when you'd want to use it. So it's a fight between two conventions: consistency with NumPy names (no underscore) and consistency with a PEP standard that prefers snake case (yes underscore).

For this, for now, though, I'll leave it. I'm prioritizing "getting stuff done" over "getting stuff perfect". It's not like we're ever going to _remove_ the `tolist` alias, such that we'd regret having promoted it in the past.

File: docs/user-guide/how-to-math-broadcasting.md (hunk 2 above)
Comment (jpivarski) [reply docs/user-guide/how-to-math-broadcasting.md]: Yes!

**Comment by jpivarski** [comment]
This PR is going to squash 6 issues at once!

**Comment by jpivarski** [comment]
That's it! No more `[todo]`!

this-is-a-stub-no-more

**Comment by jpivarski** [comment]
Even though one of the required tests is failing, packaging-tests.yml `build_cpp_wheels` on MacOS, I'm going to merge this PR anyway. The same test is failing on `main` for reasons that have nothing to do with this PR, and are probably caused by changes upstream of Awkward, anyway. Merging this PR won't make that situation any worse (or any different).

Meanwhile, merging this PR would make the new documentation go online immediately. (The new documentation doesn't describe any unreleased code changes.) That's a benefit.

But we absolutely cannot, cannot release awkward-cpp until that segfault is fixed. It will break Uproot (through AwkwardForth).

**Pull request merged**